java-checks/src/main/java/org/sonar/java/checks/security/SMTPSSLServerIdentityCheck.java

```diff
@@ -0,0 +1,139 @@
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks.security;
+
+import com.google.common.collect.ImmutableList;
+import java.util.List;
+import javax.annotation.CheckForNull;
+import org.sonar.check.Rule;
+import org.sonar.java.checks.helpers.ConstantUtils;
+import org.sonar.java.checks.methods.AbstractMethodDetection;
+import org.sonar.java.matcher.MethodMatcher;
+import org.sonar.java.matcher.TypeCriteria;
+import org.sonar.java.model.LiteralUtils;
+import org.sonar.plugins.java.api.tree.Arguments;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+@Rule(key = "S4499")
+public class SMTPSSLServerIdentityCheck extends AbstractMethodDetection {
+
+  private static final String APACHE = "org.apache.commons.mail.Email";
+  private static final String BOOLEAN = "boolean";
+  private static final String HASHTABLE = "java.util.Hashtable";
+
+  private static final MethodMatcher SET_SSL_ON_CONNECT = MethodMatcher.create()
+    .typeDefinition(TypeCriteria.is(APACHE))
+    .name("setSSLOnConnect")
+    .addParameter(BOOLEAN);
+
+  private static final MethodMatcher HASHTABLE_PUT = MethodMatcher.create()
+    .typeDefinition(TypeCriteria.is(HASHTABLE))
+    .name("put")
+    .withAnyParameters();
+
+  @Override
+  protected List<MethodMatcher> getMethodInvocationMatchers() {
+    return ImmutableList.of(SET_SSL_ON_CONNECT, HASHTABLE_PUT);
+  }
+
+  @Override
+  protected void onMethodInvocationFound(MethodInvocationTree mit) {
+    MethodTree method = findEnclosingMethod(mit);
+    if (method != null) {
+      Arguments args = mit.arguments();
+      if (SET_SSL_ON_CONNECT.matches(mit) && LiteralUtils.isTrue(args.get(0))) {
+        MethodBodyApacheVisitor apacheVisitor = new MethodBodyApacheVisitor();
+        method.accept(apacheVisitor);
+        if (!apacheVisitor.isSecured) {
+          reportIssue(mit, "Enable server identity validation on this SMTP SSL connection.");
+        }
+      } else if (HASHTABLE_PUT.matches(mit) && parametersSocketFactoryMatch(args.get(0), args.get(1))) {
```

Comment:
In the MethodMatcher `HASHTABLE_PUT` I use the function `withAnyParameters`,
because I use it twice.
In the function `onMethodInvocationFound` we proceed only if both parameters are Strings.
On the other hand, in the function `visitMethodInvocation` of the `MethodBodyHashtableVisitor`
we want the first parameter to be a String and the second boolean.
I could create a second MethodMatcher for the `MethodBodyHashtableVisitor`.